lib/src/commit_builder.rs

```diff
@@ -309,6 +325,14 @@ impl DetachedCommitBuilder {
         }
         Ok(commit)
     }
+
+    /// Writes new commit without making it visible in the repo.
+    ///
+    /// This does not consume the builder, so you can reuse the current
+    /// configuration to create another commit later.
+    pub fn write_hidden(&mut self) -> BackendResult<Commit> {
+        write_to_store(&self.store, self.commit.clone(), &self.sign_settings)
+    }
```

Comment:
> In order to render description template, we'll need a Commit object that
> represents the old state (with new tree and parents) before updating the
> commit description. The added functions will help generate an intermediate
> Commit object.
> 
> Alternatively, we can create an in-memory Commit object with some fake
> CommitId. It should be lightweight, but might cause weird issue because the
> fake id wouldn't be found in the store.
> 
> I think it's okay to write a temporary commit and rely on GC as we do for
> merge trees. However, I should note that temporary commits are more likely to
> be preserved as they are pinned by no-gc refs until "jj util gc".

@martinvonz do you think this is good idea?

The resulting `cmd_commit()` will look like:
```rust
    // set up temporary commit object
    let mut commit_builder = tx
        .mut_repo()
        .rewrite_commit(command.settings(), &commit)
        .detach();
    commit_builder.set_tree_id(tree_id);
    ...

    // render description ++ diff.summary() with the temporary commit
    // other commit keywords can be used, but branches won't work as expected
    let temp_commit = commit_builder.write_hidden()?;
    let template = description_template(&tx, None, &temp_commit, command.settings())?;
    let description = edit_description(tx.base_repo(), &template, command.settings())?

    // update description and write final commit object to store
    commit_builder.set_description(description);
    let new_commit = commit_builder.write(tx.mut_repo())?;
```

Reply:
Makes sense. I'm not worried about the extra GC-able commits.

We don't have GC at Google, but we've naturally already decide to accept the amount of trees and commits we get from regular snapshotting, so creating an extra commit object in `jj commit` should be fine. And maybe we'll implement GC some day anyway.